uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js

```diff
@@ -748,20 +762,35 @@ define([
       });
     }
 
+    /**
+     * Reloads entitlements for the currently set absence period,
+     * notifies child controllers and passes updated absence types to them
+     *
+     * @return {Promise}
+     */
+    function reloadEntitlements () {
```

Comment:
1. The name says it's reloading entitlements, but i don't see anything about loading any entitlement in this function.
2. The loading of the entitlements is in fact buried inside `setAbsenceTypesFromEntitlements`, which in its name says nothing about loading entitlements.
3. The function also emits an event about updating the balance, but why should it do that? It should be only about reloading entitlements

Overall, i suggest this refactoring (not final, of course)
```js
$rootScope.$on('LeaveRequestPopup::absencePeriodChanged', function () {
  // @NOTE when the absence period has changed, here's a clear list of things to do in succession
  reloadEntitlements()
    .then(mapAbsenceTypesWithBalance)
    .then(function () {
      $rootScope.$emit('LeaveRequestPopup::updateBalance', vm.absenceTypes);
    })
});

// @NOTE this is just about reloading entitlements now
function reloadEntitlements () {
  vm.loading.entitlements = true;

  return Entitlement.all({
    contact_id: vm.request.contact_id,
    period_id: vm.period.id,
    type_id: { IN: absenceTypesAndIds.ids }
  }, true) // `true` because we want to use the 'future' balance for calculation
  .then(function () {
    vm.loading.entitlements = false;
  });
}

// @NOTE no reason to pass `absenceTypesAndIds.types` as param, if it's already global
// @NOTE the function assigns directly the new value to `vm.absenceType`
function mapAbsenceTypesWithBalance (entitlements) {
  vm.absenceTypes = entitlements.map(function (entitlementItem) {
      var absenceType = _.find(absenceTypesAndIds.types, function (absenceTypeItem) {
```

Reply:
also, just realized: if we are *re*loading the entitlements, where are we doing it the first time?

Reply:
> also, just realized: if we are reloading the entitlements, where are we doing it the first time?

found the answer, it's done in `_loadAbsenceTypes ()`, where `setAbsenceTypesFromEntitlements` is called (again, notice how hard to follow this code is because the functions keep doing stuff that their names say nothing about)
```js
function _loadAbsenceTypes () {
  return AbsenceType.all(getAbsenceTypeParams())
     // ...
    .then(function (absenceTypes) {
      // ...
      return setAbsenceTypesFromEntitlements();
    });
```
so i would suggest to rename `reloadEntitlements` into `loadEntitlements`, to be used both on init and on period change, and that you extract the loading of the entitlements from `_loadAbsenceTypes` and put the two in a promise chain instead

Reply:
@AkA84  I attempted to do it slightly different:

Initialisation:

```js
return $q.resolve()
        .then(loadAbsenceTypes)
        .then(loadEntitlements)
        .then(setAbsenceTypesFromEntitlements)
        .then(setInitialAbsenceType)
        .then(initStatus)
        .then(initContact)
```

Reloading:

```js
$rootScope.$on('LeaveRequestPopup::absencePeriodChanged', reloadEntitlements),
// ...
function reloadEntitlements () {
      return loadEntitlements()
        .then(setAbsenceTypesFromEntitlements)
        .then(function () {
          $rootScope.$emit('LeaveRequestPopup::updateBalance', vm.absenceTypes);
        });
    }
```

With that approach you can see that each function does exactly what it is said to do. If it says "I load absence types", it really loads them only and not "loads and then does Y and Z as well"

Reply:
Still not there yet, there is inconsistency in the code. In init you do
```js
return $q.resolve()
  // ...
  .then(loadEntitlements)
  .then(setAbsenceTypesFromEntitlements)
```
so you clearly split the loading with the set-types-from-entitlements thing ("foo", to avoid repetition)

but when you reload you do
```js
$rootScope.$on('LeaveRequestPopup::absencePeriodChanged', reloadEntitlements)
```
which not only does foo inside, but it also sends an event
```js
function reloadEntitlements () {
  return loadEntitlements()
    .then(setAbsenceTypesFromEntitlements)
    .then(function () {
      $rootScope.$emit('LeaveRequestPopup::updateEntitlements', vm.absenceTypes);
    });
}
```
Why should a dev expect that `reloadEntitlements` does foo, when `loadEntitlements` doesn't? The code is inconsistent.

A couple of ideas to avoid this:

1. You create a wrapping function that loads the entitlement and does foo. You use it then both in init and in reload
2. You simply use the building blocks that you already have in the event listener handler
    ```js
    $rootScope.$on('LeaveRequestPopup::absencePeriodChanged', function () {
      loadEntitlements()
        .then(setAbsenceTypesFromEntitlements)
        .then(function () {
          $rootScope.$emit('LeaveRequestPopup::updateEntitlements', vm.absenceTypes);
        });
    })
```

Reply:
We do not want to emit an event when we don't really need it so I went with #2.